net-tools/netwatch/src/udp.rs

```diff
@@ -51,97 +68,810 @@ impl UdpSocket {
         Self::bind_raw(addr)
     }
 
+    /// Is the socket broken and needs a rebind?
+    pub fn is_broken(&self) -> bool {
+        self.is_broken.load(std::sync::atomic::Ordering::SeqCst)
+    }
+
+    /// Marks this socket as needing a rebind
+    fn mark_broken(&self) {
+        self.is_broken
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    /// Rebind the underlying socket.
+    pub fn rebind(&self) -> Result<()> {
+        debug!("rebinding {}", self.addr);
+        // Remove old socket
+        let mut guard = self.socket.write().unwrap();
+        {
+            let Some(socket) = guard.take() else {
+                bail!("cannot rebind closed socket");
+            };
+            drop(socket);
+        }
+
+        // Prepare new socket
+        let new_socket = inner_bind(self.addr)?;
+
+        // Insert new socket
+        guard.replace(new_socket);
+
+        // Clear errors
+        self.is_broken
+            .store(false, std::sync::atomic::Ordering::SeqCst);
```

Comment:
`SeqCst` will not give you anything wrong, but is overkill.  When working with simple loads and stores on bools the docs on when to use `Release` and `Acquire` are pretty straight forward: https://doc.rust-lang.org/stable/std/sync/atomic/enum.Ordering.html#variant.Release

I can't find the direct quote, but the argument by Mara Bos (who wrote Rust Atomics and Locks) is something like: SeqCst is a code-smell because it means there has been no thought about the concurrency involved.  It suggests the program needs a total order.  Or something like that.

Reply:
Agreed.
I love the "happens before relationship" mental model that mara bos explains really well in her [atomics and locks book](https://marabos.nl/atomics/memory-ordering.html).
With that mental model - what you care about is that other threads see the `is_broken` bool flip *after* the atomics in the socket lock flip.
From the perspective of the thread we're looking at, this happens-before relationship is true.
Using `Acquire` and `Release` orderings makes sure this is also the case from the perspective of other threads.

It's not *wrong* to use `SeqCst`, but mara writes something about that in the final words in the chapter:

> Myth: Sequentially consistent memory ordering is a great default and is always correct.
> [...]
> It is advisable to see SeqCst as a warning sign. Seeing it in the wild often means that either something complicated is going on, or simply that the author did not take the time to analyze their memory ordering related assumptions, both of which are reasons for extra scrutiny.

Reply:
fair, I use it to make sure I don't introduce a race condition by accident..

Reply:
fixed